README.md
# powerEQTL
R package for power analysis for eQTL study

== TOADD == 

* Below is example code using the package to generate the similar power analysis curve as GTEx study[^1]

[^1]: Lonsdale J and Thomas J, et al. The Genotype-Tissue Expression (GTEx) project. Nature Genetics, 45:580-585, 2013.


```{r}
# library(devtools);install_github("sterding/powerEQTL")  
library('powerEQTL')

# sample size
N <- c(50,100,150,200,250,300)
nn <- length(N)

# MAF
MAF <- seq(0.5,20,0.1)/100
nq <- length(MAF)

# number of SNPs tested (10 SNPs per gene x 24 genes in total)
nSNP=240

# significant level (FP)
a=0.05

# obtain power
power_unbalanced <- array(numeric(nn*nq), dim=c(nn,nq))
for (i in 1:nn){
    for (j in 1:nq){
        # unbalanced
        result <- powerEQTL.ANOVA(MAF=MAF[j], 
                                  typeI=a, 
                                  nTests=nSNP, 
                                  myntotal=N[i], 
                                  mystddev=0.13,
                                  deltaVec = c(0.13, 0.13),
                                  verbose = F)
        power_unbalanced[i,j] <-result;
    }
}

# set up graph
xrange <- range(MAF*100)
yrange <- c(0:1)
colors <- rainbow(length(N))
plot(xrange, yrange, log='x', type="n",
     xlab="MAF (%)",
     ylab="Power",
     main="Power Estimation for eQTL Studies\nSig=0.05, nSNP=240 (one-way unbalanced ANOVA)")

abline(v=0, h=seq(0,1,.1), lty=2, col="grey89")
abline(h=0, v=c(1:10), lty=2,col="grey89")

# add power curves
for (i in 1:nn){
    lines(MAF*100, power_unbalanced[i,], type="l", lwd=4, col=colors[i])
}

legend("topleft", title="Sample size (n)", as.character(N),fill=colors, bty='n')
```

NEWS
# v0.1.3 created on Jan. 24, 2017
#  add GitHub URL (https://github.com/sterding/powerEQTL) 
#   to DESCRIPTION file
# v0.1.2 created on Jan. 23, 2017
#  (1) revise "Description" section in DESCRIPTION file
#  (2) In example of "minMAFeQTL.ANOVA.Rd": "replace myntotal = 200" by "myntotal = 234"

# v0.1.1 created on Jan. 19, 2017
#  (1) fixed a typo in minMAFeQTL.sLR.Rd: value should be minimum detectable MAF
#
# v0.1.0 created on Jan. 15, 2017
#  (1) move "powerMediation" from Depends to Imports
#  (2) add functions 'minMAFeQTL.ANOVA' and 'minMAFeQTL.SLR'
# v0.0.9 created on Jan. 13, 2017
#  (1) revised 'minEffectEQTL.ANOVA'
#
# v0.0.8 created on Dec. 31, 2016
#  (1) edited online helps
#
# v0.0.7 created on Dec. 30, 2016
#  (1) added author "Scott T. Weiss"
#  (2) edited online helps
#
# v0.0.6 created on Dec. 27, 2016
#  (1) add intermediate output to 'minEffectEQTL.ANOVA'
#  (2) edited online menu for 'minEffectEQTL.ANOVA'
#
# v0.0.5 created on Dec. 26, 2016
#  (1) change depends to R (>= 3.3.0)
#  (2) change the author order
#
# v0.0.4 created on Dec. 12, 2016
#  (1) added functions 'powerEQTL.ANOVA2' and 'ssEQTL.ANOVA2'

# v0.0.3 created on Dec. 12, 2016
#  (1) fixed a bug in powerEQTL.ANOVA.R: 
#  gm1 = deltaVec[1]
#  should be
#  gm1 = -deltaVec[1]
#
#  (2) fixed a bug in minEffectEQTL.ANOVA: need to directly use simplified power calculation formula. So add function 'powerEQTL.ANOVA'
#
# v0.0.2 created on Dec. 11, 2016
#  (1) relax the assumption in ANOVA approach: mu2-mu1 can be different from mu3-mu2
#  (2) add function minEffectEQTL.ANOVA to calculate minimum detectable effect size
#      delta/sigma. This would be helpful in design stage where we have no prior
#      info about delta and sigma. delta=mu2-mu1=mu3-mu2.


R/ssEQTL.SLR.R
# created on Dec. 8, 2016
#  (1) sample size calculation for eQTL analysis based on simple linear regression
#

# MAF - minor allele frequency
# typeI - type I error rate
# nTests - number of tests
# myntotal - total number of subjects
# mypower - desired power
# mystddev - standard deviation of gene expression levels
#            (assume each group of subjects has the same mystddev)
# n.lower - lower bound for sample size
# n.upper - upper bound for sample size
# verbose - flag indicating if intermedaite results should be output
ssEQTL.SLR=function(MAF,
                       typeI=0.05,
                       nTests=200000,
                       slope=0.13,
                       mypower=0.8,
                       mystddev=0.13,
                       n.lower = 2.01,
                       n.upper = 1e+30,
                       verbose=TRUE)
{

  sigma.x=sqrt(2*MAF*(1-MAF))
  alpha = typeI/nTests

  aa=ss.SLR(power=mypower,
               lambda.a=slope,
               sigma.x=sigma.x,
               sigma.y=mystddev,
               n.lower = n.lower,
               n.upper = n.upper,
               alpha = alpha,
               verbose = verbose)
  return(aa$n)

}



R/ssEQTL.ANOVA.R
# created on Dec. 11, 2016
# (1) allow mu2-mu1 not equal to mu3-mu2,
#   where mu1 is the mean expression level for mutation homozygote,
# mu2 is the mean expression level for heterozygote,
#  and mu3 is the mean expression level for wildtype homozygote
#
# created on Dec. 8, 2016
#  (1) sample size calculation for eQTL analysis based on ANOVA or simple linear regression
#

# squared difference between estimated power and desired power
diffPowerFunc.ANOVA=function(myntotal,
                       MAF=0.1,
                       typeI=0.05,
                       nTests=200000,
                       mystddev=0.13,
                       deltaVec=c(0.13,0.13),
                       desiredPower=0.8)
{
  estPower=powerEQTL.ANOVA(MAF=MAF,
                              typeI=typeI,
                              nTests=nTests,
                              myntotal=myntotal,
                              mystddev=mystddev,
                              deltaVec=deltaVec, verbose=FALSE)
  diff=(estPower-desiredPower)^2
  return(diff)

}

# MAF - minor allele frequency
# typeI - type I error rate
# nTests - number of tests
# mypower - desired power
# mystddev - standard deviation of gene expression levels
#            (assume each group of subjects has the same mystddev)
# deltaVec - mean difference of gene expression levels between groups
#            deltaVec[1]=mu2-mu1 and deltaVec[2]=mu3-mu2
ssEQTL.ANOVA=function(MAF,
                   typeI=0.05,
                   nTests=200000,
                   mypower=0.8,
                   mystddev=0.13,
                   deltaVec=c(0.13,0.13))
{
  res.optim=optim(par=200, fn=diffPowerFunc.ANOVA,typeI=typeI,
                  nTests=nTests, mystddev=mystddev,
                  deltaVec=deltaVec, desiredPower=mypower,
        method = c("L-BFGS-B"),
        lower = 1, upper = 1000000)
  myntotal=res.optim$par

  return(myntotal)
}



diffPowerFunc.ANOVA3=function(
                              myntotal,
                              MAF=0.1,
                              typeI=0.05,
                              nTests=200000,
                              effsize=1,
                              desiredPower=0.8)
{
  estPower=powerEQTL.ANOVA2(effsize=effsize,
                            MAF=MAF,
                            typeI=typeI,
                            nTests=nTests,
                            myntotal=myntotal,
                            verbose=FALSE)
  diff=(estPower-desiredPower)^2
  return(diff)
  
}


ssEQTL.ANOVA2=function(
  effsize,
  MAF,
  typeI=0.05,
  nTests=200000,
  mypower=0.8
)
{
  res.optim=optim(par=200, fn=diffPowerFunc.ANOVA3,
                  MAF=MAF,
                  typeI=typeI,
                  nTests=nTests, 
                  effsize=effsize,
                  desiredPower=mypower,
                  method = c("L-BFGS-B"),
                  lower = 1, upper = 1000000)
  myntotal=res.optim$par
  
  return(myntotal)
}




R/powerEQTL.SLR.R
# created on Dec. 8, 2016
#  (1) power calculation for eQTL analysis based on simple linear regression
#

# MAF - minor allele frequency
# typeI - type I error rate
# nTests - number of tests
# myntotal - total number of subjects
# mypower - desired power
# mystddev - standard deviation of gene expression levels
#            (assume each group of subjects has the same mystddev)

powerEQTL.SLR=function(MAF,
                          typeI=0.05,
                          nTests=200000,
                          slope=0.13,
                          myntotal=200,
                          mystddev=0.13,
                          verbose=TRUE)
{

  sigma.x=sqrt(2*MAF*(1-MAF))
  alpha = typeI/nTests

  aa=power.SLR(n=myntotal,
                   lambda.a=slope,
                   sigma.x=sigma.x,
                   sigma.y=mystddev,
                   alpha = alpha,
                   verbose = verbose)
  return(aa$power)

}



R/powerEQTL.ANOVA.R
# created on Dec. 11, 2016
# (1) allow mu2-mu1 not equal to mu3-mu2,
#   where mu1 is the mean expression level for mutation homozygote,
# mu2 is the mean expression level for heterozygote,
#  and mu3 is the mean expression level for wildtype homozygote
#
# created on Dec. 8, 2016
#  (1) power calculation for eQTL analysis based on ANOVA or simple linear regression
#

# MAF - minor allele frequency
# typeI - type I error rate
# nTests - number of tests
# myntotal - total number of subjects
# mystddev - standard deviation of gene expression levels
#            (assume each group of subjects has the same mystddev)
# deltaVec - mean difference of gene expression levels between groups
#            deltaVec[1]=mu2-mu1 and deltaVec[2]=mu3-mu2
# verbose - flag indicating if we should output intermediate results

powerEQTL.ANOVA=function(MAF,
                   typeI=0.05,
                   nTests=200000,
                   myntotal=200,
                   mystddev=0.13,
                   deltaVec=c(0.13, 0.13),
                   verbose=TRUE)
{
  if(length(deltaVec)!=2)
  {
    stop("'deltaVec' has 2 and only 2 elements")
  }
  
  gm1 = -deltaVec[1]
  gm2 = 0
  gm3 = deltaVec[2]

  n2=MAF^2
  n1=2*MAF*(1-MAF)
  n0=(1-MAF)^2

  w2=1
  w1=n1/n2
  w0=n0/n2

  alpha=typeI/nTests

  if(verbose)
  {
    cat("mu2=", gm1, ", mu1=", gm2, ", mu0=", gm3, "\n")
    cat("n2=", n2, ", n1=", n1, ", n0=", n0, "\n")
    cat("alpha=", alpha, "\n")
  }

  k=3
  mydf1=k-1
  mydf2=myntotal-k
  q=qf(p=1-alpha, df1=mydf1, df2=mydf2)

  wVec=c(w2, w1, w0)
  wVec=wVec/sum(wVec, na.rm=TRUE)
  muVec=c(gm1, gm2, gm3)

  mu=sum(wVec*muVec, na.rm=TRUE)

  myncp = myntotal*sum(wVec*(muVec-mu)^2, na.rm=TRUE)
  myncp=myncp/(mystddev^2)

  power=1-pf(q=q, df1=mydf1, df2=mydf2,
             ncp=myncp)

  return(power)

}


# effsize=delta/sigma
powerEQTL.ANOVA2=function(effsize,
                          MAF,
                          typeI=0.05,
                          nTests=200000,
                          myntotal=200,
                          verbose=TRUE)
{

  myncp=2*MAF*(1-MAF)*myntotal*effsize^2

  alpha=typeI/nTests
  
  if(verbose)
  {
    cat("myncp=", myncp, "\n")
    cat("alpha=", alpha, "\n")
  }
  
  k=3
  mydf1=k-1
  mydf2=myntotal-k
  q=qf(p=1-alpha, df1=mydf1, df2=mydf2)
  
  power=1-pf(q=q, df1=mydf1, df2=mydf2,
             ncp=myncp)
  
  return(power)
  
}



R/minMAFeQTL.ANOVA.R
# created on Jan. 15, 2017
# calculate MAF based on simple linear regression
#

diffPowerFunc.ANOVA.MAF=function(MAF,
                              effsize,
                              myntotal,
                              typeI=0.05,
                              nTests=200000,
                              desiredPower=0.8)
{
  estPower=powerEQTL.ANOVA2(effsize=effsize,
                            MAF=MAF,
                           typeI=typeI,
                           nTests=nTests,
                           myntotal=myntotal,
                           verbose=FALSE)
  diff=(estPower-desiredPower)
  return(diff)

}


# MAF - minor allele frequency
# typeI - type I error rate
# nTests - number of tests
# myntotal - total number of subjects
# mystddev - standard deviation of gene expression levels
#            (assume each group of subjects has the same mystddev)
# deltaVec - mean difference of gene expression levels between groups
#            deltaVec[1]=mu2-mu1 and deltaVec[2]=mu3-mu2
# verbose - flag indicating if we should output intermediate results

minMAFeQTL.ANOVA=function(effsize, 
                   typeI=0.05,
                   nTests=200000,
                   myntotal=200,
                   mypower = 0.8,
                   verbose=TRUE)
{
  res.uniroot=uniroot(f=diffPowerFunc.ANOVA.MAF,
                  interval=c(0.000001, 0.5),
                  effsize = effsize,
                  myntotal=myntotal,
                  typeI=typeI,
                  nTests=nTests,  
                  desiredPower=mypower
                  )
  
  if(verbose)
  {
    cat("Results of uniroot>>>\n")
    print(res.uniroot)
  }
  MAF=res.uniroot$root
  
  return(MAF)
}



R/minSlopeEQTL.SLR.R
# created on Dec. 8, 2016
#  (1) minimum slope calculation for eQTL analysis based on simple linear regression
#

# MAF - minor allele frequency
# typeI - type I error rate
# nTests - number of tests
# myntotal - total number of subjects
# mypower - desired power
# mystddev - standard deviation of gene expression levels
#            (assume each group of subjects has the same mystddev)
# verbose - flag indicating if intermedaite results should be output
minSlopeEQTL.SLR=function(MAF,
                   typeI=0.05,
                   nTests=200000,
                   myntotal=200,
                   mypower=0.8,
                   mystddev=0.13,
                   verbose=TRUE)
{

  sigma.x=sqrt(2*MAF*(1-MAF))
  alpha = typeI/nTests

  aa=minEffect.SLR(n=myntotal,
                   power=mypower,
                   sigma.x=sigma.x,
                   sigma.y=mystddev,
                   alpha = alpha,
                   verbose = verbose)
  return(aa$lambda.a)

}



R/minEffectEQTL.ANOVA.R
# created on Dec. 11, 2016
# calculate minimum detectable effect size (delta/sigma) given sample size and power
#
# created on Dec. 8, 2016
#  (1) power calculation for eQTL analysis based on ANOVA or simple linear regression
#

diffPowerFunc.ANOVA2=function(effsize,
                              myntotal,
                              MAF=0.1,
                              typeI=0.05,
                              nTests=200000,
                              desiredPower=0.8)
{
  estPower=powerEQTL.ANOVA2(effsize=effsize,
                            MAF=MAF,
                           typeI=typeI,
                           nTests=nTests,
                           myntotal=myntotal,
                           verbose=FALSE)
  diff=(estPower-desiredPower)
  return(diff)

}


# MAF - minor allele frequency
# typeI - type I error rate
# nTests - number of tests
# myntotal - total number of subjects
# mystddev - standard deviation of gene expression levels
#            (assume each group of subjects has the same mystddev)
# deltaVec - mean difference of gene expression levels between groups
#            deltaVec[1]=mu2-mu1 and deltaVec[2]=mu3-mu2
# verbose - flag indicating if we should output intermediate results

minEffectEQTL.ANOVA=function(MAF, 
                   typeI=0.05,
                   nTests=200000,
                   myntotal=200,
                   mypower = 0.8,
                   verbose=TRUE)
{
  res.uniroot=uniroot(f=diffPowerFunc.ANOVA2,
                  interval=c(0.0001, 1000),
                  myntotal=myntotal,
                  typeI=typeI,
                  MAF = MAF,
                  nTests=nTests,  
                  desiredPower=mypower
                  )
  
  if(verbose)
  {
    cat("Results of uniroot>>>\n")
    print(res.uniroot)
  }
  effsize=res.uniroot$root
  
  return(effsize)
}



R/minMAFeQTL.SLR.R
# created on Jan. 15, 2017
#  (1) minimum MAF calculation for eQTL analysis based on simple linear regression
#



diffPowerFunc.SLR=function(MAF,
                           slope,   
                           myntotal,
                           mystddev=0.13,
                              typeI=0.05,
                              nTests=200000,
                              desiredPower=0.8)
{
  estPower=powerEQTL.SLR(MAF=MAF,
                            typeI=typeI,
                            nTests=nTests,
                            slope = slope,
                            myntotal=myntotal,
                            mystddev=mystddev,
                            verbose=FALSE)
  diff=(estPower-desiredPower)
  return(diff)
  
}




# slope - slope of the simple linear regression
# typeI - type I error rate
# nTests - number of tests
# myntotal - total number of subjects
# mypower - desired power
# mystddev - standard deviation of gene expression levels
#            (assume each group of subjects has the same mystddev)
# verbose - flag indicating if intermedaite results should be output
minMAFeQTL.SLR=function(slope,
                   typeI=0.05,
                   nTests=200000,
                   myntotal=200,
                   mypower=0.8,
                   mystddev=0.13,
                   verbose=TRUE)
{

  res.uniroot=uniroot(f=diffPowerFunc.SLR,
    interval=c(0.000001, 0.5),
    slope=slope,   
    myntotal=myntotal,
    mystddev=mystddev,
    typeI=typeI,
    nTests=nTests,
    desiredPower=mypower)
                        
  if(verbose)
  {
    cat("Results of uniroot>>>\n")
    print(res.uniroot)
  }
  MAF=res.uniroot$root
  
  return(MAF)
}

